### PR TITLE
testing the projects zome init function with mockHDK 

### LINKED ISSUE #29: write a test for projects -> init
This one should be easier than others to start, which I am basing off of the notion of 'total HDK function calls'

First, add `pub` to it so you can reference it outside the crate, in the tests folder. 
Create a new test file that corresponds to this `lib.rs` file, over in `tests` crate folder.
Writing the test will require drilling into the two function calls, then looking for calls to the HDK that will need to be mocked

https://github.com/h-be/acorn/blob/7fab3552fad60036eca4dd8f8d5b2aebbdd45139/dna/zomes/projects/src/lib.rs#L17-L27

### PR DESCRIPTION
**Pull request by WFinck97**
closes #29

### CODE REVIEW

File: dna/tests/src/project/project_init.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                access: ().into(),
+                functions,
+            }),
+        );
+        let header_hash = fixt!(HeaderHash);
+        mock_hdk
+            .expect_create()
+            .with(mockall::predicate::eq(expected))
+            .times(1)
+            .return_const(Ok(header_hash.clone()));
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:45]: great! you even got the `.with` which requires mocking a suitable match for the input type. It does strengthen the nature of the test

File: dna/tests/src/project/project_init.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        
+        // init also calls join_project_during_init, drilling down, it calls the following hdk functions under the hood
+        // TODO: handle cases where a path with 1 or more parents doesn't yet exist in the DHT (this would call `create` and `create_link` additional times)
+        let entry_hash = fixt!(EntryHash);
+        mock_hdk
+            .expect_hash_entry()
+            .times(3)
+            .return_const(Ok(entry_hash.clone()));
+
+        let expected_output = fixt!(Element);
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:55]:
```suggestion
        // relates to `Path::from(MEMBER_PATH).ensure()?;`
        let expected_output = fixt!(Element);
```

File: dna/tests/src/project/project_init.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        let exp_vec = vec![Some(expected_output)];
+        mock_hdk
+            .expect_get()
+            .times(1)
+            .return_const(ExternResult::Ok(exp_vec));
+
+        let agent_info = fixt!(AgentInfo);
+
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:63]:
```suggestion
```

File: dna/tests/src/project/project_init.rs (hunk 2 above)
Comment (Connoropolous) [reply dna/tests/src/project/project_init.rs:55]: thinking of adding comments like this, when the actual HDK call has been nested

File: dna/tests/src/project/project_init.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        let header_hash = fixt!(HeaderHash);
+        mock_hdk
+            .expect_create()
+            .with(mockall::predicate::eq(expected))
+            .times(1)
+            .return_const(Ok(header_hash.clone()));
+        
+        // init also calls join_project_during_init, drilling down, it calls the following hdk functions under the hood
+        // TODO: handle cases where a path with 1 or more parents doesn't yet exist in the DHT (this would call `create` and `create_link` additional times)
+        let entry_hash = fixt!(EntryHash);
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:49]:
```suggestion
        // 1. `Path::from(MEMBER_PATH).ensure()?;`
        // 2. `let member_path_address = Path::from(MEMBER_PATH).hash()?;`
        // 3. `let member_entry_hash = hash_entry(&member)?;`
        let entry_hash = fixt!(EntryHash);
```

given these three different calls, we could set up the mock to handle each unique one, which I think might be good. 
Now that we're advancing from your first pass, which is excellent, to the second we can do this. Using `.with` and making sure that arguments consumed are expected, and then returning relevant results

File: dna/tests/src/project/project_init.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        
+        mock_hdk
+            .expect_create_link()
+            .times(1)
+            .return_const(Ok(header_hash));
+
+        set_hdk(mock_hdk);
+
+        let result = init(());
+        assert_eq!(result.is_ok(), true);
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:82]: There is an alternative we can use here, called `assert_matches`, and match against the whole result

https://docs.rs/assert_matches/1.5.0/assert_matches/macro.assert_matches.html

File: dna/tests/src/project/project_init.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    use projects::project::project_meta::validate::*;
+    use projects::init;
+
+    #[test]
+    fn test_init() {
+               
+        let mut mock_hdk = MockHdkT::new();
+        
+        let zome_info = fixt!(ZomeInfo);
+        // do these mock functions need to be called in the same order as the actual code?
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:21]: good question, we might have to try to break one sometime to find out

File: dna/tests/src/project/project_init.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    use holochain_types::prelude::ElementFixturator;
+    use holochain_types::prelude::ValidateDataFixturator;
+    use projects::project::error::Error;
+    use projects::project::project_meta::crud::ProjectMeta;
+    use projects::project::project_meta::validate::*;
+    use projects::init;
+
+    #[test]
+    fn test_init() {
+
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:17]: let's add a comment here that describes the overall behaviour of `init` that we're testing, i.e. the fact that it makes two main function calls, and most importantly the purpose of those function calls, focused on being human readable

File: dna/tests/src/project/project_init.rs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        
+        mock_hdk
+            .expect_create()
+            .times(1)
+            .return_const(Ok(header_hash.clone()));
+        
+        mock_hdk
+            .expect_create_link()
+            .times(1)
+            .return_const(Ok(header_hash));
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:77]: right on! 
now as we move to second pass you can add `.with` to these and also specify distinct `header_hash` values as the return values, to keep the test more realistic

**Review by Connoropolous — CHANGES_REQUESTED** [review]
Great work, you have clearly got the general idea, and I think the hang of it. 

We can step up/into creating the realistic .with values and realistic return_const values. 

Think about giving meaningful variables names to your `fixt!` calls, as we mock, so that it makes it easy to track them, and use them in the right contexts.

File: dna/tests/src/project/project_init.rs (hunk 1 above)
Comment (WFinck97) [reply dna/tests/src/project/project_init.rs:45]: I actually found a test function for `create_receive_signal_cap_grant` in your crud_hdk crate, and just copied that.

Comment (Connoropolous) [reply dna/tests/src/project/project_init.rs:51]: that's totally cool. Resourcefulness is key

File: dna/tests/src/project/project_init.rs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                access: ().into(),
+                functions,
+            }),
+        );
+        let header_hash = fixt!(HeaderHash);
+        mock_hdk
+            .expect_create()
+            .with(mockall::predicate::eq(expected))
+            .times(1)
+            .return_const(Ok(header_hash.clone()));
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:51]: to make this modular, and reusable in the future, let's set up a function with a signature like this (it returns the unit type `()` which we can omit which is quivalent): 
```
fn setup_create_receive_signal_cap_grant_mock(mock_hdk: MockHdkT) {
  
}
```

move the code highlighted in to there

File: dna/tests/src/project/project_init.rs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        let create_link_input = CreateLinkInput::new(
+            member_path_entry_hash,
+            member_hash,
+            LinkTag::from(()),
+        );
+        mock_hdk
+            .expect_create_link()
+            .with(mockall::predicate::eq(create_link_input))
+            .times(1)
+            .return_const(Ok(header_hash));
```
Comment (Connoropolous) [inline dna/tests/src/project/project_init.rs:114]: do the same with this one.

Then, within the main test, call those two functions, passing in the `mock_hdk`

**Review by Connoropolous — CHANGES_REQUESTED** [review]
Sweet, so in the future, I will leave more of my full review comments all at once, but it has been useful this time to do it in passes. The thing which stands out right now is that we may want to reuse that code, while also making our code more readable, so I've suggested a way to do that. Everything you've done looks on point! 🚀

File: dna/tests/src/project/project_init.rs (hunk 1 above)
Comment (WFinck97) [reply dna/tests/src/project/project_init.rs:51]: :wink:

**Review by Connoropolous — APPROVED** [review]

**Pull request merged**